crates/bevy_render/src/mesh/mesh/mod.rs

```diff
@@ -553,18 +554,56 @@ impl Mesh {
             .as_float3()
             .expect("`Mesh::ATTRIBUTE_POSITION` vertex attributes should be of type `float3`");
 
-        let normals: Vec<_> = positions
-            .chunks_exact(3)
-            .map(|p| face_normal(p[0], p[1], p[2]))
-            .flat_map(|normal| [normal; 3])
-            .collect();
+        match self.indices() {
+            Some(indices) => {
+                let mut count: usize = 0;
+                let mut corners = [0_usize; 3];
+                let mut normals = vec![[0.0f32; 3]; positions.len()];
+                let mut adjacency_counts = vec![0_usize; positions.len()];
+
+                for i in indices.iter() {
+                    corners[count % 3] = i;
+                    count += 1;
+                    if count % 3 == 0 {
+                        let normal = face_normal(
+                            positions[corners[0]],
+                            positions[corners[1]],
+                            positions[corners[2]],
+                        );
+                        for corner in corners {
+                            normals[corner] =
+                                (Vec3::from(normal) + Vec3::from(normals[corner])).into();
+                            adjacency_counts[corner] += 1;
+                        }
+                    }
+                }
+
+                // average (smooth) normals for shared vertices...
+                // TODO: support different methods of weighting the average
+                for i in 0..normals.len() {
+                    let count = adjacency_counts[i];
+                    if count > 0 {
+                        normals[i] = (Vec3::from(normals[i]) / (count as f32)).normalize().into();
+                    }
+                }
```

Comment:
```suggestion
                let mut normals = vec![[0.0; 3]; positions.len()];
                let mut adjacency_counts = vec![0; positions.len()];

                for face in indices.chunks(3) {
                    let normal = face_normal(positions[face[0]], positions[face[1]], positions[face[2]]);
                    for &index in &face {
                        normals[index as usize] = (Vec3A::from(normal) + Vec3A::from(normals[corner])).into();
                        adjacency_counts[index as usize] += 1;
                    }
                }

                // Average (smooth) normals for shared vertices.
                // TODO: Weight by angle.
                for (normal, &count) in normals.iter_mut().zip(adjacency_counts.iter()) {
                    if count > 0 {
                        *normal = (Vec3A::from(*normal) / count as f32).normalize().into();
                    }
                }
```

Reply:
> for face in indices.chunks(3) {

There does not seem to be a method called chunks in Indices and I do not know how to take 3 at a time in some other way.

Reply:
`chunks` is a method on any iterator: https://doc.rust-lang.org/std/primitive.slice.html

should probably use `chunks_exact` here though